yuyo/modals.py

```diff
@@ -0,0 +1,1440 @@
+        ------
+        ValueError
+            If the custom_id is already registered.
+        """
+        if custom_id in self._prefix_ids:
+            raise ValueError(f"{custom_id!r} is already registered as a prefix match")
+
+        if custom_id in self._modals:
+            raise ValueError(f"{custom_id!r} is already registered as a normal match")
+
+        if timeout is NO_DEFAULT:
+            timeout = BasicTimeout(datetime.timedelta(10), max_uses=1)
+
+        elif timeout is None:
+            timeout = NeverTimeout()
+
+        if prefix_match:
+            self._prefix_ids[custom_id] = (timeout, modal)
+            return self
+
+        self._modals[custom_id] = (timeout, modal)
+        return self
+
+    def get_modal(self, custom_id: str, /) -> typing.Optional[AbstractModal]:
+        """Get the modal set for a custom ID.
+
+        Parameters
+        ----------
+        custom_id
+            The custom_id to get the modal for.
+
+        Returns
+        -------
+        AbstractModal | None
+            The callback for the custom_id, or [None][] if it doesn't exist.
+        """
+        if entry := self._modals.get(custom_id) or self._prefix_ids.get(custom_id):
+            return entry[1]
+
+        return None
+
+    def remove_modal(self, custom_id: str, /) -> Self:
+        """Remove the modal set for a custom ID.
+
+        Parameters
+        ----------
+        custom_id
+            The custom_id to unset the modal for.
+
+        Returns
+        -------
+        Self
+            The modal client to allow chaining.
+
+        Raises
+        ------
+        KeyError
+            If the custom_id is not registered.
+        """
+        try:
+            del self._modals[custom_id]
+        except KeyError:
+            del self._prefix_ids[custom_id]
+
+        return self
+
+
+class AbstractModal(abc.ABC):
+    """Base class for a modal execution handler."""
+
+    __slots__ = ()
+
+    @abc.abstractmethod
+    async def execute(self, ctx: ModalContext, /) -> None:
+        """Execute this modal.
+
+        Parameters
+        ----------
+        ctx
+            The context to execute this with.
+        """
+
+
+class _TrackedField:
+    __slots__ = ("custom_id", "default", "parameter", "prefix_match", "type")
+
+    def __init__(
+        self,
+        *,
+        custom_id: str,
+        default: typing.Union[typing.Any, NoDefault],
+        parameter: str,
+        prefix_match: bool,
+        type_: hikari.ComponentType,
+    ) -> None:
+        self.custom_id = custom_id
+        self.default = default
+        self.parameter = parameter
+        self.prefix_match = prefix_match
+        self.type = type_
+
+
+class Modal(AbstractModal, typing.Generic[_CallbackSigT]):
+    """Standard implementation of a modal executor.
+
+    Examples
+    --------
+
+    There's a few different ways this can be used to for a modal.
+
+    Sub-components can be added to an instance of a modal using chainable
```

Comment:
Say how to send the components also this can be phrased like ActionColumnExecutor's docs